mopro-ffi/src/halo2/mod.rs

```diff
@@ -1,69 +1,172 @@
+            std::collections::HashMap<String, mopro_ffi::Halo2VerifyFn>,
+        > = once_cell::sync::Lazy::new(|| set_halo2_verifying_circuits());
+
+        fn generate_halo2_proof(
+            in0: String,
+            in1: String,
+            in2: std::collections::HashMap<String, Vec<String>>,
+        ) -> Result<mopro_ffi::GenerateProofResult, mopro_ffi::MoproError> {
+            let name = std::path::Path::new(in1.as_str()).file_name().unwrap();
+            if let Some(prove_fn) = HALO2_PROVING_CIRCUITS.get(name.to_str().unwrap()) {
+                prove_fn(&in0, &in1, in2)
+            } else {
+                Err(mopro_ffi::MoproError::Halo2Error(
+                    format!(
+                        "Unknown Prove Circuit: {}. Have Prove Circuits: {:?}",
+                        in1,
+                        HALO2_PROVING_CIRCUITS.keys()
+                    )
+                    .to_string(),
+                ))
+            }
+        }
+
+        fn verify_halo2_proof(
+            in0: String,
+            in1: String,
+            in2: Vec<u8>,
+            in3: Vec<u8>,
+        ) -> Result<bool, mopro_ffi::MoproError> {
+            let name = std::path::Path::new(in1.as_str()).file_name().unwrap();
+            if let Some(verify_fn) = HALO2_VERIFYING_CIRCUITS.get(name.to_str().unwrap()) {
+                verify_fn(&in0, &in1, in2, in3)
+            } else {
+                Err(mopro_ffi::MoproError::Halo2Error(
+                    format!(
+                        "Unknown Verify Circuit: {}. Have Verify Circuits: {:?}",
+                        in1,
+                        HALO2_VERIFYING_CIRCUITS.keys()
+                    )
+                    .to_string(),
+                ))
+            }
+        }
+    };
 }
 
-pub fn verify_halo2_proof(proof: Vec<u8>, public_inputs: Vec<u8>) -> Result<bool, MoproError> {
-    panic!("stub implementation: TODO: copy logic from mopro-core to here");
-    // let deserialized_proof: halo2::SerializableProof =
-    //     bincode::deserialize(&proof).map_err(|e| MoproError::Halo2Error(e.to_string()))?;
-    // let deserialized_inputs: halo2::SerializablePublicInputs =
-    //     bincode::deserialize(&public_inputs).map_err(|e| MoproError::Halo2Error(e.to_string()))?;
-    // let is_valid = halo2::verify_halo2_proof(deserialized_proof, deserialized_inputs).unwrap();
-    // Ok(is_valid)
+/// Proving Circuits are Halo2 Circuits that can generate proofs
+/// Provide the circuits that you want to be able to generate proofs for
+/// as a list of pairs of the form `circuit_proving_key`, `prove_fn`
+/// Where `circuit_proving_key` is the name of the proving key file
+/// and `prove_fn` is the function that generates the proof available locally
+/// NOTE: YOU CAN ONLY USE THIS MACROS ONCE IN YOUR CODEBASE, IN THE SAME MODULE AS `app!()`
+/// NOTE: TO USE THIS MACRO, YOU MUST HAVE THE `mopro-ffi/halo2` FEATURE ENABLED
+#[macro_export]
+macro_rules! set_halo2_proving_circuits {
```

Comment:
Same here, remove this macro and just let the user write a normal function with a certain name mapping inputs to proof functions.

Reply:
Same thoughts as above